### PR TITLE
Enforce the `offline_bot` fixture in `Test*WithoutRequest`

### PR DESCRIPTION
**Pull request by Bibo-Joshi**
Follow-up for #4454. based on that branch, merge after.
Uses the `offline_bot` in all the `*WithoutRequest` tests and fixes the very few places where requests where still being made.
The check for enforcing `offline_bot` is now configured to run only on GitHub and only in one of the several test matrix runs. IMO this should suffice and avoids running additional if-checks on every single test function unnecessarily.

### CODE REVIEW

File: tests/test_bot.py [hunk 1]
```diff
@@ -732,7 +735,7 @@ async def make_assertion(url, request_data: RequestData, *args, **kwargs):
                         "id": "123",
                         "type": "document",
                         "document_url": (
-                            "https://raw.githubusercontent.com/python-telegram-bot"
+                            "https://raw.githubusercontent.com/python-telegram-offline_bot"
```
Comment (harshil21) [inline tests/test_bot.py:738]:
```suggestion
                            "https://raw.githubusercontent.com/python-telegram-bot"
```

File: tests/test_bot.py [hunk 2]
```diff
@@ -750,7 +753,7 @@ async def make_assertion(url, request_data: RequestData, *args, **kwargs):
             else:
                 button_dict = {
                     "text": "button_text",
-                    "web_app": {"url": "https://python-telegram-bot.org"},
+                    "web_app": {"url": "https://python-telegram-offline_bot.org"},
```
Comment (harshil21) [inline tests/test_bot.py:756]:
```suggestion
                    "web_app": {"url": "https://python-telegram-bot.org"},
```
offline ptb would be interesting

File: tests/test_bot.py [hunk 3]
```diff
@@ -763,7 +766,7 @@ async def make_assertion(url, request_data: RequestData, *args, **kwargs):
             InlineQueryResultDocument(
                 id="123",
                 document_url=(
-                    "https://raw.githubusercontent.com/python-telegram-bot/logos/master/"
+                    "https://raw.githubusercontent.com/python-telegram-offline_bot/logos/master/"
```
Comment (harshil21) [inline tests/test_bot.py:769]:
```suggestion
                    "https://raw.githubusercontent.com/python-telegram-bot/logos/master/"
```

File: tests/test_bot.py [hunk 4]
```diff
@@ -779,15 +782,15 @@ async def make_assertion(url, request_data: RequestData, *args, **kwargs):
             )
         elif button_type == "web_app":
             button = InlineQueryResultsButton(
-                text="button_text", web_app=WebAppInfo("https://python-telegram-bot.org")
+                text="button_text", web_app=WebAppInfo("https://python-telegram-offline_bot.org")
```
Comment (harshil21) [inline tests/test_bot.py:785]:
```suggestion
                text="button_text", web_app=WebAppInfo("https://python-telegram-bot.org")
```

File: tests/test_bot.py [hunk 5]
```diff
@@ -840,7 +843,7 @@ async def make_assertion(url, request_data: RequestData, *args, **kwargs):
                         "type": "document",
                         "document_url": (
                             "https://raw.githubusercontent.com/"
-                            "python-telegram-bot/logos/master/logo/png/"
+                            "python-telegram-offline_bot/logos/master/logo/png/"
```
Comment (harshil21) [inline tests/test_bot.py:846]:
```suggestion
                            "python-telegram-bot/logos/master/logo/png/"
```

File: tests/test_bot.py [hunk 6]
```diff
@@ -853,14 +856,14 @@ async def make_assertion(url, request_data: RequestData, *args, **kwargs):
...
-        monkeypatch.setattr(bot.request, "post", make_assertion)
+        monkeypatch.setattr(offline_bot.request, "post", make_assertion)
         results = [
             InlineQueryResultArticle("11", "first", InputTextMessageContent("first")),
             InlineQueryResultArticle("12", "second", InputMessageContentLPO("second")),
             InlineQueryResultDocument(
                 id="123",
                 document_url=(
-                    "https://raw.githubusercontent.com/python-telegram-bot/logos/master/"
+                    "https://raw.githubusercontent.com/python-telegram-offline_bot/logos/master/"
```
Comment (harshil21) [inline tests/test_bot.py:866]:
```suggestion
                    "https://raw.githubusercontent.com/python-telegram-bot/logos/master/"
```

File: tests/test_bot.py [hunk 7]
```diff
@@ -932,7 +935,7 @@ async def make_assertion(url, request_data: RequestData, *args, **kwargs):
                         "type": InlineQueryResultType.DOCUMENT,
                         "document_url": (
                             "https://raw.githubusercontent.com/"
-                            "python-telegram-bot/logos/master/logo/png/"
+                            "python-telegram-offline_bot/logos/master/logo/png/"
```
Comment (harshil21) [inline tests/test_bot.py:938]:
```suggestion
                            "python-telegram-bot/logos/master/logo/png/"
```

File: tests/test_bot.py [hunk 8]
```diff
@@ -959,7 +962,7 @@ async def make_assertion(url, request_data: RequestData, *args, **kwargs):
             InlineQueryResultDocument(
                 id="123",
                 document_url=(
-                    "https://raw.githubusercontent.com/python-telegram-bot/logos/master/"
+                    "https://raw.githubusercontent.com/python-telegram-offline_bot/logos/master/"
```
Comment (harshil21) [inline tests/test_bot.py:965]:
```suggestion
                    "https://raw.githubusercontent.com/python-telegram-bot/logos/master/"
```

File: tests/test_bot.py [hunk 9]
```diff
@@ -1713,7 +1729,7 @@ async def make_assertion(
         try:
             replace_button = InlineKeyboardButton(text="replace", callback_data="replace_test")
             no_replace_button = InlineKeyboardButton(
-                text="no_replace", url="http://python-telegram-bot.org/"
+                text="no_replace", url="http://python-telegram-offline_bot.org/"
```
Comment (harshil21) [inline tests/test_bot.py:1732]:
```suggestion
                text="no_replace", url="http://python-telegram-bot.org/"
```

File: tests/test_bot.py [hunk 10]
```diff
@@ -1722,24 +1738,25 @@ async def make_assertion(
...
+            offline_bot.username
+            monkeypatch.setattr(offline_bot, "_post", make_assertion)
             results = [
                 InlineQueryResultArticle(
                     "11", "first", InputTextMessageContent("first"), reply_markup=reply_markup
                 ),
                 InlineQueryResultVoice(
                     "22",
-                    "https://python-telegram-bot.org/static/testfiles/telegram.ogg",
+                    "https://python-telegram-offline_bot.org/static/testfiles/telegram.ogg",
```
Comment (harshil21) [inline tests/test_bot.py:1750]:
```suggestion
                    "https://python-telegram-bot.org/static/testfiles/telegram.ogg",
```

File: tests/test_keyboardbutton.py [hunk 11]
```diff
@@ -81,7 +81,7 @@ def test_to_dict(self, keyboard_button):
         assert keyboard_button_dict["request_users"] == keyboard_button.request_users.to_dict()
 
     @pytest.mark.parametrize("request_user", [True, False])
-    def test_de_json(self, bot, request_user):
+    def test_de_json(self, offline_bot, request_user):
```
Comment (harshil21) [inline tests/test_keyboardbutton.py:84]:
```suggestion
    def test_de_json(self, request_user):
```
unused arg it seems

File: tests/conftest.py [hunk 12]
```diff
@@ -98,6 +98,34 @@ def pytest_collection_modifyitems(items: List[pytest.Item]):
...
+    # let's not slow down the tests too much with these additional checks
+    # that's why we run them only in GitHub actions and only on *one* of the several test
+    # matrix entries
+    @pytest.fixture(autouse=True)
+    def _disallow_requests_in_without_request_tests(request):
+        """This fixture prevents tests that don't require requests from using the online-bot."""
+
+        if type(request).__name__ == "SubRequest":
+            # We want to allow *WithoutRequest test classes to use fixtures that do use requests,
+            # e.g. `animation` and so on. Unfortunately the `SubRequest` class is not public, so
```
[the remark this answers is not in the document]
Comment (Bibo-Joshi) [reply tests/conftest.py:111]: We can't really 100% enforce no-network - you can still build a `Bot` manually or directly use httpx. The only way to properly do that is to deactivate httpx/sockets interely and in #4317 I found no sane way to do that.
I see this PR as sane-effort approach on trying to keep requests out of `*WithoutRequest`. TBH I personally still don't really use that distinction at all 😅

[the remark this answers is not in the document]
Comment (harshil21) [reply tests/conftest.py:111]: imo the comment should say that instead of saying we *want* to

[the remark this answers is not in the document]
Comment (Bibo-Joshi) [reply tests/conftest.py:111]: Fair enough. I'll find a wording

**Review by harshil21 — APPROVED** [review]
LGTM!

**Pull request merged**